subprojects/core-api/src/main/java/org/gradle/api/provider/Property.java

```diff
@@ -34,30 +43,40 @@
  */
 public interface Property<T> extends Provider<T>, HasConfigurableValue {
     /**
-     * Sets the value of the property the given value, replacing whatever value the property already had.
+     * Sets the value of the property to the given value, replacing whatever value the property already had.
      *
-     * <p>This method can also be used to discard the value of the property, by passing {@code null} as the value.
-     * The convention for this property, if any, will be used to provide the value instead.
+     * <p>This method can also be used to discard the value of the property, by passing it {@code null}.
+     * When the value is discarded (or has never been set in the first place), the convention (default value)
+     * for this property, if specified, will be used to provide the value instead.</p>
      *
      * @param value The value, can be null.
      */
     void set(@Nullable T value);
 
     /**
-     * Sets the property to have the same value of the given provider, replacing whatever value the property already had. This property will track the value of the provider and query its value each time the value of the property is queried. When the provider has no value, this property will also have no value.
+     * Sets the property to have the same value as the given provider, replacing whatever value the property already had.
+     * This property will track the value of the provider and query its value each time the value of the property is queried.
+     * When the provider has no value, this property will also have no value.
      *
-     * <p>
-     * When the given provider represents a task output, this property will also carry the task dependency information from the provider.
-     * </p>
+     * <p>This method can NOT be used to discard the value of the property. Specifying a {@code null} provider will result
```

Comment:
What happens if the set provider (non-null) returns `null` as its value? Will the property then use the convention value (if set)?

Reply:
I would interpret this javadoc to mean that after setting Property A to use another Provider B, if when A is queried, B returns null, then A's value is null, NOT whatever A's convention is.

I'd update this to "When the provider has no value, this property will also have no value **- regardless of whether or not a convention has been set.**"

Reply:
Tom is right, that is what happens. When the provider has no value, the property will have no value either, regardless of the convention being set or not. That doesn't sound right to me, but we're documenting here, not changing behaviour. Have added a test to confirm this. Will also look into what happens when the convention is set to a provider without value.